File: noteifier.js
const env = require('node-env-file')(__dirname + '/.env');
const { checkFiles, contains } = require('./handlers/fileReader');
const { sendNotification } = require('./handlers/notifier');
const { execute, open, openManyFiles } = require('./handlers/execution');
const { lookupProcess } = require('./handlers/processes');

const { DESIRED, appToKeyword, appToName } = require('./models/constants');

const pathToDocs = __dirname + '/Documents/';

/*
[X] for each of the processes in the list
[X]   check processes
[X]   if the processes you're looking for are newly opened*
[X]     read through the files to check for keywords
[X]     if the file has the keyword add it to the list of files
[X]       if there are any files, send the notification
[X]         if the notification is clicked
[X]           open the files
*/

let accountedFor = [];

//initial load of all open applications
DESIRED.forEach((application) => {
  lookupProcess(application, function() {
    accountedFor.push(application);
  })
})

//main process on an interval loop
setInterval( function() {
  DESIRED.forEach((application) => {
    if (!accountedFor.includes(application)) {
      lookupProcess(application, function() {
        const name = appToName(application);
        const files = checkFiles(file => contains(file, appToKeyword(application)));
        if (files) {
          accountedFor.push(application);
          sendNotification('Noteifier', `Click on me to open your notes about ${name}!`, files, openManyFiles);
        }
      })
    }
  })
}, 10000)

File: noteifier.py
from models.constants import monitored_applications, monitored_applications2
from handlers.processes import parse_application_names, get_active_applications
from handlers.tools import check_files, generate_open_files_command, notify
import threading
import time


class Noteifier():
    def __init__(self):
        self.accounted_for = {app for app in parse_application_names(get_active_applications()) if app in monitored_applications2}
        self.paused = threading.Event()
        self.paused.clear()

    def run(self):
        while True:
            while not self.paused.is_set():
                applications_open = get_active_applications()
                open_application_names = parse_application_names(applications_open)
                for application in monitored_applications2:
                    if application in open_application_names and application not in self.accounted_for:
                        files = check_files(application)
                        if files:
                            cmd = generate_open_files_command(files)
                            time.sleep(2)
                            notify(message='Click on me to open notes about {}'.format(application), title='Noteifier', command=cmd)
                        self.accounted_for.add(application)

    def pause(self):
        self.paused.set()

    def resume(self):
        self.paused.clear()
